### PR TITLE
Add autodocs

### PR DESCRIPTION
**Pull request by JAB1305**
This adds the necessary configuration files for automatically generating documentation based on  the code docstrings, similarly to how it works in the [BaSyx-Python-SDK Docs](https://github.com/eclipse-basyx/basyx-python-sdk/tree/main/docs). 

We also add a simple GitHub workflow CI job to verify the compilation of docs without errors. 

Depends on #5

### CODE REVIEW

File: docs/add-requirements.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Additional requirements for building the docs
+sphinx~=7.2
+sphinx-rtd-theme~=2.0
+sphinx-argparse~=0.4.0
+sphinx_autodoc_typehints~=2.2.3
+basyx-python-sdk~=1.0.0
```
Comment (s-heppner) [inline docs/add-requirements.txt]: Is this the correct dependency? Shouldn't it be https://github.com/aas-core-works/aas-core3.0-python?
Or what is this needed for?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: ci
```
Comment (s-heppner) [inline .github/workflows/ci.yml]: I think the best workflow for this would be to close #5 first and then merge `main` into this branch. 
That way, you can adapt the correct CI configuration.

File: .readthedocs.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Read the Docs configuration file
+# See https://docs.readthedocs.io/en/stable/config-file/v2.html for details
+
+version: 2
+
+build:
+  os: ubuntu-20.04
```
Comment (s-heppner) [inline .readthedocs.yaml]: I think current version of Ubuntu is 24.04, so we should probably update this.
If there's the option to specify to always use the latest version, then this might be preferable.

File: docs/source/conf.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'Eclipse BaSyx Python Framework'
+project_copyright = str(datetime.datetime.now().year) + ', the Eclipse BaSyx Authors'
+author = 'The Eclipse BaSyx Authors'
+
+# The full version, including alpha/beta/rc tags
+release = "none"
```
Comment (s-heppner) [inline docs/source/conf.py:35]: I'm wondering if there is the option to extract this number from the `pyproject.toml`, rather than having to change this manually with every new release.

File: docs/source/conf.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+autodoc_default_options = {
+    'members': True,
+    'undoc-members': True
+}
+
+# Mapping for correctly linking other module documentations.
+intersphinx_mapping = {
+    'python': ('https://docs.python.org/3', None),
+    'aas_core3': ('https://aas-core30-python.readthedocs.io/en/latest/', None),
+    'basyx-python-sdk': ('https://basyx-python-sdk.readthedocs.io/en/latest/', None)
```
Comment (s-heppner) [inline docs/source/conf.py]: Where do we need this?

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+lxml>=4.2,<5
+python-dateutil>=2.8,<3.0
+types-python-dateutil
+pyecma376-2>=0.2.4
+urllib3>=1.26,<2.0
+Werkzeug>=3.0.3,<4
+schemathesis~=3.7
+hypothesis~=6.13
+lxml-stubs~=0.5.1
+aas-core3.0~=1.0.4
```
Comment (s-heppner) [inline requirements.txt]: I think we should not introduce the requirement.txt. Instead the requirements are specified in the more modern pyproject.toml. 
Is this file really needed for sphinx? If so, then maybe we need to change something in the sphinx configuration.

File: sdk/basyx/object_store.py [hunk 7]
```diff
@@ -6,48 +6,55 @@
 # SPDX-License-Identifier: MIT
 """
 This module implements Registries for the AAS, in order to enable resolving global
-`Identifiers`; and mapping `Identifiers` to `Identifiable` objects.
+:attr:`~basyx.aas.model.base.Identifier` and mapping
```
Comment (s-heppner) [inline sdk/basyx/object_store.py]: This should be `aas_core3.types.Identifier` and so on, since that's the classes that are used.
Also, can you double check the `:attr:` designation? I think `Identifier should be `:class:` or something similar.

File: sdk/basyx/object_store.py [hunk 8]
```diff
@@ -6,48 +6,55 @@
...
 """
 
 import abc
 from typing import MutableSet, Iterator, Generic, TypeVar, Dict, List, Optional, Iterable
 
 from aas_core3.types import Identifiable, Referable, Class
 
+""" TODO: Add docstring """
 _IdentifiableType = TypeVar('_IdentifiableType', bound=Identifiable)
+""" TODO: Add docstring """
```
Comment (s-heppner) [inline sdk/basyx/object_store.py]: I think an inline comment at line 18 should suffice:

```
# We define types for :class:`aas_core3.types.Identifier` and :class:`aas_core3.types.Referable` for easier referencing.
```

**Review by s-heppner — CHANGES_REQUESTED** [review]

File: .readthedocs.yaml (hunk 3 above)
Comment (JAB1305) [reply .readthedocs.yaml]: Yes, 24.04 would be the neweset version. I changed it to "ubuntu-lts-latest" so it should always use the latest LTS available on RTD' side.

File: requirements.txt (hunk 6 above)
Comment (JAB1305) [reply requirements.txt]: This file is not directly necessary for the sphinx compilation but rather a way to install all required packages in the CI-jobs context. You are right, we should definetely achieve this by using the pyproject.toml and adding everything .
However there is also currently the 'add-requirements.txt' which is only used by the docs job to install the additional dependencies necessary for compilation. This one I would keep as its a pretty neat way to add all the imports.
I also just noticed that i still have to check wether every of these entries is acutally necessary.

Comment (JAB1305) [reply requirements.txt]: Okay I just rechecked and it is indeed necessary to have at least one requirements.txt file that is passed to readthedocs for the compilation of the sphinx docs on their side.
We could definetely achieve a structure only based on the pyproject.toml by defining an additional, optional, dependency set like so:
```
[project.optional-dependencies]
docs = [
  "sphinx",
  "sphinx-rtd-theme",
  "sphinx-argparse",
  "sphinx_autodoc_typehints",
  "basyx-python-sdk"
] 
```
and adjusting the imports used in the github CI. But I don't see a way this could be passed to RTD https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html.

I think our best option would be to drop the toplevel 'requirements.txt' and only keep one; containing the dependency AND additional dependencies.

But then we would have to declare dependencies in the pyproject.toml as well as in the requirements file which I am not really a fan of. Maybe we could generate the requirements file to always display whatever is declared in the pyproject.toml dependencies AND optional dependencies? Or are we just fine with declaring every dependency at two different points?

File: sdk/basyx/object_store.py (hunk 7 above)
Comment (JAB1305) [reply sdk/basyx/object_store.py]: About the `:attr:` vs `:class:`:
I somehow locked in on repairing these docstrings, still containing the outdated basyx-python-sdk classes. In trying to achieve proper links via intersphinx I used a tool to generate every single entry currently documentated by the old documentation. By looking through that list I found out that some entries are `:attr:` type, others `:class:`.
For exmaple Identifier only exists as `:attr:`, using it with `:class:` results in the reference not being found and therefore no intersphinx link to be placed.
I attached the file in case you want to look at it yourself. 
[all_docs.txt](https://github.com/user-attachments/files/16922044/all_docs.txt)

File: docs/source/conf.py (hunk 4 above)
Comment (JAB1305) [reply docs/source/conf.py:35]: I found a way to extract the release from the pyproject.toml. Maybe you can recheck what I did there

File: sdk/basyx/object_store.py (hunk 7 above)
Comment (JAB1305) [reply sdk/basyx/object_store.py]: I updated the docstrings to use the aas_core3 classes now

File: docs/source/conf.py (hunk 5 above)
Comment (JAB1305) [reply docs/source/conf.py]: The intersphinx mappings in general are necessary to automatically link to the docs listed here when refering to an aas_core3 or python class.
The basyx-python-sdk entry is not necessary anymore, I removed it.

File: docs/add-requirements.txt (hunk 1 above)
Comment (JAB1305) [reply docs/add-requirements.txt]: This is also an artifact of the old basyx-python-sdk where i copied the basic structure from, hence now removed.

Comment (JAB1305) [reply docs/add-requirements.txt]: aas-core3.0 is not listed here, as its not an additional requirement for doc compilation. Its already a dependency of the project itself

Comment (s-heppner) [reply docs/add-requirements.txt]: I understand, makes sense.

File: requirements.txt (hunk 6 above)
Comment (s-heppner) [reply requirements.txt]: Let's talk about this in person when we next meet.

Comment (JAB1305) [reply requirements.txt]: Okay so as discussed in our meeting I restructured the dependency specifications.
Unfortunately I was not able to verify RTD compilation so I cant tell wether the current setup would function and wether every necessary dependency is listed. Maybe this will have to be changed later on (again).

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.X_PYTHON_VERSION }}
+    - name: Install Python dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r docs/docs-requirements.txt
+    - name: Check documentation for errors
+      run: |
+        SPHINXOPTS="-a -E -n -W --keep-going" make -C docs html
```
Comment (s-heppner) [inline .github/workflows/ci.yml]: Can you add an empty line at the end of the file please?

**Review by s-heppner — CHANGES_REQUESTED** [review]
Other than the small comment above, it lgtm now.
While you're at it, can you resolve the conflicts with the `ci.yml` in the main branch?

File: .github/workflows/ci.yml (hunk 9 above)
Comment (s-heppner) [reply .github/workflows/ci.yml]: I think this file also has conflicts with the current state of the main branch

Comment (JAB1305) [reply .github/workflows/ci.yml]: Resolving the conflicts was a bit painful as I basically overwrote the ci.yml file in my branch. I rebased onto main and force pushed whilst removing some commits. I rechecked and the results look good. It can also be merged without conflicts now.

**Review by s-heppner — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 9 above)
Comment (s-heppner) [reply .github/workflows/ci.yml]: Thank you for your hassle!
